Infer public webtransport addrs from quic-v1 addrs.

MarcoPolo opened the pull request:
[edited]

Adds a pure function to infer public webtransport addrs from quic addrs if we see that we are using the same port for both quic and webtransport addrs.

For example say we know we are listening on:

- `/ip4/0.0.0.0/udp/1234/quic-v1`
- `/ip4/0.0.0.0/udp/1234/quic-v1/webtransport`

And we observe that our public IP addresses include
- `/ip4/2.3.4.5/udp/1234/quic-v1`

We see that we're listening on the same address for quic-v1 and webtransport, so we should be able to say pretty confidently that we are also reachable via `/ip4/2.3.4.5/udp/1234/quic-v1/webtransport`.

That's what this PR does.

marten-seemann reviewed (changes requested):
Strongly opposed to this PR. While I appreciate the problem it’s trying to solve, I don’t think we should solve it in this way.

I’ve started early work on the address pipeline, and part of that is moving cruft _out_ of the basic host (speaking of, can I get a review on https://github.com/libp2p/go-libp2p/pull/2248 please?). This PR adds more technical debt to the basic host, and will make implementing the address pipeline harder.

MarcoPolo commented:
Summarizing sync convo:

We both agreed the previous approach was too hacky. We also agreed there was value in having this fix out so that we can start seeing webtransport addrs in the wild. We came up with an alternate approach that makes this a pure function that is pretty well contained. When we fix this with the address pipeline, users won't notice the difference. Things will just keep working.

I've updated the PR to use the less hacky approach, and added some table tests. I've also verified this surfaces my public webtransport addrs on Kubo.

marten-seemann commented on p2p/host/basic/basic_host.go:
```diff
@@ -997,10 +998,83 @@ func (h *BasicHost) AllAddrs() []ma.Multiaddr {
 
+var wtComponent = ma.StringCast("/webtransport")
+
+// inferWebtransportAddrsFromQuic infers more webtransport addresses from QUIC addresses.
+// This is useful when we discover our public QUIC address, but haven't discovered our public WebTransport addrs.
+// If we see that we are listening on the same port for QUIC and WebTransport,
+// we can be pretty sure that the WebTransport addr will be reachable if the
+// QUIC one is.
```
Great explanation, I wish every function we have was that well documented!

marten-seemann commented on p2p/host/basic/basic_host.go:
```diff
@@ -997,10 +998,83 @@ func (h *BasicHost) AllAddrs() []ma.Multiaddr {
+	// We need to check if we are listening on the same ip+port for QUIC and WebTransport.
+	// If not, there's nothing to do since we can't infer anything.
+
+	// Count the number of QUIC addrs, this will let us allocate just once at the beginning.
```
Not sure if the plan to reduce allocs actually works out, `SplitLast` also allocates pretty heavily. No need to change it though, as this is temporary anyway.

MarcoPolo replied:
Does it? I thought it just pointed to slices?

marten-seemann commented on p2p/host/basic/basic_host.go:
```diff
@@ -997,10 +998,83 @@ func (h *BasicHost) AllAddrs() []ma.Multiaddr {
+
+		if _, lastComponent := ma.SplitLast(addr); lastComponent.Protocol().Code == ma.P_QUIC_V1 {
+			addrStr := addr.String()
+			if _, ok := quicOrWebtransportAddrs[addrStr]; ok {
```
If I understand correctly, this will make the function misbehave if the input contains duplicates. We should probably also have a test case that covers that case.

MarcoPolo replied:
I'll dedup

marten-seemann replied:
Resolved by running deduping on the input of this function.

marten-seemann commented on p2p/host/basic/basic_host.go:
```diff
@@ -997,10 +998,83 @@ func (h *BasicHost) AllAddrs() []ma.Multiaddr {
+	out := make([]ma.Multiaddr, 0, len(in)+(quicAddrCount-len(webtransportAddrs)))
+	for _, addr := range in {
+		// Add all the original addresses
+		out = append(out, addr)
```
Why not just append to `in`?

MarcoPolo replied:
I didn't want to mutate the in list in case someone else kept a reference to it (A similar bug bit me with a related PR in this area of the code).

MarcoPolo replied:
Also if we append to in, the function is no longer pure

marten-seemann replied:
Kind of? The caller's slice wouldn't change if we're only appending, we'd only be modifying slice elements between `len` and `cap`. But I'm fine either way, this will heavily change with the new multiaddr design anyway.

marten-seemann reviewed (commented):
I'm struggling a bit to wrap my head around the logic here. I think there's a bug here if the input contains duplicates.

By the way, the code would be sooo much cleaner with the new multiaddr API.
